Adding Tutorials for Segmentation and Linking

snilsn opened the pull request:
## Description

* additional notebook on the parameters of Segmentation
* additional notebook on the parameters of Linking
* bugfix in feature detection part 1 (actually use _weighted_abs_ as `position_threshold`)
* setting a random seed for the added noise in feature detection part 1

I worked on these notebooks for a while now, hence the long commit history. Comparing Jupyter notebooks seems to cause some trouble for git.

## Checklist for Requirements

### General
* [x] Have you followed our guidelines in Section "Contributing" from README.md? 

### Style
* [x] Have you checked already existing notebooks to get an idea about our tutorial style?
* [x] Have you extensively used markdown elements to structure and explain your notebook?
* [x] Did you include your tutorial notebook in the `index.rst` file?

### Data & Reproducibility
* [ ] Have you uploaded your example data to [zenodo]( https://zenodo.org/ ) (see README) if needed?
   - does not apply

### Concerning Tests
* [x] Have you self-reviewed your code / notebook and corrected any misspellings? 
* [x] Have you checked if your notebook is running in the recommended standard conda enviroment [see here]((docs/RunningTutorials.md) and updates the [list of required python packages](./requirements.txt) if needed?
* [x] Have you carefully checked that sphinx-based rendering is still working correctly?

### Git Aspects: To Make It Easier for Us!
* [x] Have you written descriptive commit messages? 
* [ ] Have you kept your pull request small and limited so that it is easy to review? 
    - This probably isn't considered small
* [x] Have the newest changes been merged into this branch?

snilsn committed:
started first notebook

snilsn committed:
animation of track created

snilsn committed:
better explanations, tracking described

snilsn committed:
plots improved

snilsn committed:
analysis added

snilsn committed:
features notebook started

snilsn committed:
assymetric dataset, position_threshold visualised

snilsn committed:
cleaning up outputs

snilsn committed:
sigma threshold visual

snilsn committed:
sigma threshold explanation

snilsn committed:
trying min_distance

snilsn committed:
n_min_threshold explained

snilsn committed:
erosion demonstrated

snilsn committed:
new folder + restructuring

snilsn committed:
cells deleted

snilsn committed:
better graphs

snilsn committed:
cleaned up Features_2

snilsn committed:
wavelength filtering reference

snilsn committed:
dublicate deleted

snilsn committed:
segmentation notebook started

snilsn committed:
segmentation notebook started

snilsn committed:
first plots

snilsn committed:
first plots

snilsn committed:
Segmentation threshold visualized

snilsn committed:
Segmentation threshold visualized

snilsn committed:
Merge pull request #1 from snilsn/master

merge changes of main

snilsn committed:
3d data explored

snilsn committed:
3d data explored

snilsn committed:
segmentation finished, new notebooks created

snilsn committed:
segmentation finished, new notebooks created

snilsn committed:
explored the linking parameters

snilsn committed:
explored the linking parameters

snilsn committed:
more parameters explained, animation added

snilsn committed:
more parameters explained, animation added

snilsn committed:
cleanup 3d, startet analysis

snilsn committed:
cleanup 3d, startet analysis

snilsn committed:
adaptive search and subnet_size explained

snilsn committed:
adaptive search and subnet_size explained

snilsn committed:
cleaned up Linking notebook

snilsn committed:
cleaned up Linking notebook

snilsn committed:
3d data finished

snilsn committed:
3d data finished

snilsn committed:
notebook with splitting/merging dataset

snilsn committed:
notebook with splitting/merging dataset

snilsn committed:
Merge pull request #2 from fsenf/idealized-cases

Idealized cases

snilsn committed:
Merge branch 'tobac-project:master' into master

snilsn committed:
delete dublicate

snilsn committed:
delete dublicate

snilsn committed:
change style of segmentation

snilsn committed:
resolve conflicts

snilsn committed:
remove advanced notebooks

snilsn committed:
Merge branch 'tobac-project:master' into tutorials

snilsn committed:
resolve conflicts

snilsn committed:
clean up structure

snilsn committed:
revision of segmentation started

snilsn committed:
better graphs, more explanation, memory parameter

snilsn committed:
finishing linking, renaming

snilsn committed:
formatting with black

snilsn committed:
bugfixing weighted_abs

snilsn committed:
adding random seed

snilsn committed:
updating index.rst

snilsn committed:
clear outputs

snilsn commented:
I need to clean up this PR, there are many changes in the notebooks that don't have any purpose. I'll try to revise this soon.

JuliaKukulies commented:
OK @snilsn, let us know when you are ready for the review!

snilsn committed:
clean up jsons

snilsn commented:
I cleaned most of the jupyter leftovers. The diffs in feature detection part 1 are still somewhat messy, but the relevant changes I have done there are in line **792** and **969**.

This is now ready for review

JuliaKukulies commented on tutorials/Basics/Methods-and-Parameters-for-Linking.ipynb:
```diff
@@ -0,0 +1,1455 @@
+   "id": "9c3ada63",
+   "metadata": {},
+   "source": [
+    "Since we know that this dataset contains 3 features, whe can visualize the found tracks with masks created from the *cell* variable:"
```
```suggestion
    "Since we know that this dataset contains 3 features, we can visualize the found tracks with masks created from the *cell* variable:"
```

JuliaKukulies commented on tutorials/Basics/Methods-and-Parameters-for-Linking.ipynb:
```diff
@@ -0,0 +1,1455 @@
+   "id": "f9ac562e",
+   "metadata": {},
+   "source": [
+    "The starting point of the cell index if of course arbitrary. If we want to change it from 1 to 0 we can do this with the `cell_number_start` parameter:"
```
```suggestion
    "The starting point of the cell index is of course arbitrary. If we want to change it from 1 to 0 we can do this with the `cell_number_start` parameter:"
```

JuliaKukulies commented on tutorials/Basics/Methods-and-Parameters-for-Linking.ipynb:
```diff
@@ -0,0 +1,1455 @@
+    "r_{max} = v_{max} \\Delta t.\n",
+    "$$\n",
+    "\n",
+    "By reducing the maximum speed fromm 100 m/s to 70 m/s we will find more cells, because a feature that is not assigned to an existing track will be used as a starting point for a new one:"
```
```suggestion
    "By reducing the maximum speed from 100 m/s to 70 m/s we will find more cells, because a feature that is not assigned to an existing track will be used as a starting point for a new one:"
```

JuliaKukulies commented on tutorials/Basics/Methods-and-Parameters-for-Linking.ipynb:
```diff
@@ -0,0 +1,1455 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA7QAAAGGCAYAAABL6s0hAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAC9n0lEQVR4nOz9eZxkV3nY/3+ee+vW3nvP0rNLM2IkELLkLxAcPGhsBF9HttnMEltGSIZfRJyEODKW5MhIIINtSCTLgO1gcCKwg4MEQoSfMd6QYjm2WGRhFkkIpJnRrD3Te1fXfu/z/eNW91TXdFdXVVd1dc88b736NX3rnHvuqVb17Tp1nvMcUVWMMcYYY4wxxpiNxul2B4wxxhhjjDH ...
```
This goes back to an earlier discussion we had I think: Do you know what defines the best feature of all possible features within the search radius? Rather the distance or rather the direction?

snilsn replied:
Yes, I remember. In the meantime I read a bit in the trackpy documentation and in the paper about the Crocker-Grier algorithm.

As I understand it:
As long as we set the method in `tobac ` to _random_, the distance to the original position is relevant. The velocity in successive frames is assumed to be uncorrelated.
If we change the keyword to _predict_, `tobac` uses the _NearestVelocityPredict_ class of `trackpy` and guesses a new position of the cells based on the velocity in the last frame. The distance from this guessed position is then determining the linking. 

In this sense, it is in any case a distance that is decisive. But I think it would be a good idea to create a notebook of an idealized case exploring this, also for my own understanding. So I will start working on this.

JuliaKukulies replied:
Thank you for this clarification @snilsn ! That helped a lot and how great that you plan to create another notebook to make this also more clear!

JuliaKukulies commented on tutorials/Basics/Methods-and-Parameters-for-Linking.ipynb:
```diff
@@ -0,0 +1,1455 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA7QAAAGGCAYAAABL6s0hAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAC9n0lEQVR4nOz9eZxkV3nY/3+ee+vW3nvP0rNLM2IkELLkLxAcPGhsBF9HttnMEltGSIZfRJyEODKW5MhIIINtSCTLgO1gcCKwg4MEQoSfMd6QYjm2WGRhFkkIpJnRrD3Te1fXfu/z/eNW91TXdFdXVVd1dc88b736NX3rnHvuqVb17Tp1nvMcUVWMMcYYY4wxxpiNxul2B4wxxhhjjDH ...
```
iterativly -> iteratively 
adaptiv -> adaptive

JuliaKukulies reviewed (approved):
Just a few typos, but otherwise approved from my side. Again - super nice and helpful notebooks @snilsn !

snilsn committed:
fixing typos

snilsn commented:
Thanks for your feedback @JuliaKukulies. The typos are fixed now and I will start working on the notebook I mentioned.

JuliaKukulies commented:
Perfect @snilsn ! I will go ahead and merge these new notebooks after the approval of @fsenf

fsenf commented:
@snilsn : The bugfix in `tutorials/Basics/Methods-and-Parameters-for-Feature-Detection_Part_1.ipynb` looks good! 

I started to look at `tutorials/Basics/Methods-and-Parameters-for-Linking.ipynb` and wondered why the orange cell track 1 splits up into three parts when limiting the max. speed to 70 m/s. If I understand the method `tobac.testing.make_sample_data_2D_3blobs_inv()` right, all three cells have speeds between 26 and 32 m/s. Do I get something wrong?

fsenf commented:
Ahh, sorry! Forget my comment above: Speed of the 2nd and 3rd cell is linearily increasing with time 

```python
track2_x = x_0_2 + 24 * (t_temp * 60) ** 2 / 1000
```
etc.

fsenf commented:
for Linking Notebook:

The adaptive method still works with
```python
tp.linking.Linker.MAX_SUB_NET_SIZE_ADAPTIVE = 0
```
This is strange ....

fsenf committed:
updated Linking Notebook
- added some velocity calculations
- bugfix: needed to rotated the 3blob data - x & y were exchanged by accident

fsenf commented:
@snilsn : 
- **linking**: Now, I finished my work on the linking notebook. I found some plotting bugs which I resolved in the last commit. Please look carefully through my changes.
- **segmentation**: has the same plotting bug in it. Could you please resolve the bug by your own. I will do the review of the notebook thereafter.

Thanks!

snilsn commented:
Thanks for your feedback @fsenf, I will work in the changes for the segmentation notebook probably tomorrow. 

> for Linking Notebook:
> 
> The adaptive method still works with
> 
> ```python
> tp.linking.Linker.MAX_SUB_NET_SIZE_ADAPTIVE = 0
> ```
> 
> This is strange ....

After a little bit of research I came to the conclusion that this is probably the result of this line:
https://github.com/soft-matter/trackpy/blob/00defed6d6daa310c17b7e5fd28fd748394671bc/trackpy/linking/subnetlinker.py#L445

The `adaptive_step` is multiplied until only one backwards and one forward candidate is left. This line then just assigns these to one track without checking for an oversized subnet. Kinda makes sense, because why would I set a maximum lower than 1. Maybe I can add an exception for values of `subnetwork_size` lower than 1 to this [PR](https://github.com/tobac-project/tobac/pull/168) ( which btw solves all this mess with having to declare `tp.linking.Linker.MAX_SUB_NET_SIZE_ADAPTIVE` outside of the function)

snilsn committed:
Fixing plotting and explanation of subnets

snilsn committed:
formatting with black

snilsn commented:
@fsenf The changes in the plots are done now. I also fixed some typos I found and the explanation of the search radius.

fsenf committed:
three small spelling corrections

fsenf commented:
@snilsn : I corrected a few spelling mistakes I found. Now, I am happy. Thanks for this contributions!
